src/common/device_helpers.cuh

```diff
@@ -872,14 +878,43 @@ class AllReducer {
 #ifdef XGBOOST_USE_NCCL
     /** \brief this >monitor . init. */
     this->device_ordinals = device_ordinals;
-    comms.resize(device_ordinals.size());
-    dh::safe_nccl(ncclCommInitAll(comms.data(),
-                                  static_cast<int>(device_ordinals.size()),
-                                  device_ordinals.data()));
-    streams.resize(device_ordinals.size());
+    this->device_counts.resize(rabit::GetWorldSize());
+    this->comms.resize(device_ordinals.size());
+    this->streams.resize(device_ordinals.size());
+    this->id = GetUniqueId();
```

Comment:
This initialisation code looks good to me at first glance. We will need to check it explicitly with a number of test cases however.

Reply:
My thought for test cases would be to update `param` in the [basic distributed python test](https://github.com/dmlc/xgboost/blob/master/tests/distributed/test_basic.py#L12). It could be replaced by `param[xgb.rabit.get_rank()]` ... namely, parameters are a function of the process rank ID. I could assign GPU resources based on rank.

I do need to some information about the test distributed system, though.